### ISSUE TITLE
Ranger enters busy-loop if STDIN file descriptor is closed

### ISSUE THREAD

**Issue by igoyak** [ranger/ranger#1367]
I hope this will solve an old heisenbug. It doesn't seem to always reproduce naturally, but can be provoked using gdb. 

Potentially related, unresolved bug reports: 
- https://github.com/ranger/ranger/issues/87
- https://github.com/ranger/ranger/issues/525

#### Runtime Environment
- Operating system and version: Ubuntu 18.04.1 LTS
- Terminal emulator and version: rxvt-unicode (urxvt) v9.22 - released: 2016-01-23
- ranger version: ranger-master v1.9.2-53-ge56c74bb
- Python version: 2.7.15rc1 (default, Apr 15 2018, 21:51:34) [GCC 7.3.0]
- Locale: None.None
- Shell: GNU bash, version 4.4.19(1)-release (x86_64-pc-linux-gnu)

#### Current Behavior

Ranger enters a busy-loop if STDIN is closed. 

#### Expected Behavior

Ranger exits if STDIN is closed. 

#### Background

Using ranger extensively, I occasionally notice that a ranger process is using 100% of the CPU. 

It is not uncommon for me to start ranger, navigate to a directory, spawn a shell to run a command and restart ranger again. 
I also usually close terminal windows using the window manager, rather than exiting each subprocess running in that window. 

#### Analysis

To the best of my understanding, this is how the bug manifests. 

When a shell exits (e.g. when closing a terminal), it sends a SIGHUP signal to its child processes. 
If the child process does not explicitly handle this signal (which ranger doesn't do), the default behavior is to terminate the process. 

When ranger spawns a shell which restarts ranger, the SIGHUP reaches the first ranger process, but not (always?) the second process (due to the fork-exec?). The terminal and the first ranger process exit, and the remaining ranger process continues to run, but becomes orphaned and adopted by init (pid 1). 

Ranger runs a 'while True' loop which repeatedly wait for user input, which happens here: https://github.com/ranger/ranger/blob/e56c74bb30ad2f363280e4d1a14bf2c20dff5b5f/ranger/gui/ui.py#L239

	key = self.win.getch()

https://docs.python.org/2/library/curses.html#curses.window.getch

If there is no input, getch() waits for some time, then returns -1. However, because the STDIN file descriptor was closed by the earlier exited processes, getch() presumably throws an error and returns -1 immediately, which causes ranger to enter a continuous busy-loop. 

#### Possible Solutions

If getch() returns -1 and os.isatty(0) returns False, ranger should exit. 

##### Natural way (not guaranteed to work)

In a terminal, start ranger: 

  > ranger

Press 'S' to spawn a shell
Start ranger again:

  > ranger

(Maybe press down a couple of times)
Close the terminal (e.g. Alt-F4)

The former ranger process exits, but the latter enters a busy-loop. 

##### gdb way

Force-close the STDIN file descriptor of a running ranger process

Start ranger
Attach to process with gdb: 

  > gdb -p PID_OF_RANGER

Close fd 0:

  > (gdb) p close(0)

Detach to let ranger keep running:

  > (gdb) detach

Ranger enters busy-loop.

**Comment by SuperShinyEyes** [comment]
I had the exact same issue when it showed 100% of my CPU on htop.

**Comment by pascalwhoop** [comment]
I'll follow up on this one as I regulary have to `pkill -f ranger` because it clogs up my CPU cores

**Comment by igoyak** [comment]
I have spent some time developing a cross-platform way to reproduce this using Docker. 


Save as `Dockerfile`: 

	FROM ubuntu:xenial-20190610

	RUN apt-get update
	RUN apt-get install -y git
	RUN apt-get install -y python3
	RUN apt-get install -y python
	RUN apt-get install -y expect-dev
	RUN apt-get install -y gdb

	RUN git clone https://github.com/ranger/ranger.git

	WORKDIR ranger

	# Checkout a modern commit, comment out to use latest commit
	RUN git checkout 3e27a680f82f759eebd90a20a4920d4984413d6b

	ADD run /

	CMD /run

	# How to run this docker container: 
	# docker build . -t rtest; docker run --rm -ti --cap-add=SYS_PTRACE --security-opt seccomp=unconfined rtest
	# The security flags are needed to allow gdb to attach to running processes. 


Save as `run`:
	
	#!/bin/bash

	# This script should be run from inside Docker

	# Start ranger using unbuffer to imitate an interative terminal
	unbuffer ./ranger.py & 
	rangerpid=$!

	sleep 1

	reset
	rangerpid="$(pgrep ranger)"
	echo "ranger pid: $rangerpid"

	# Use gdb to attach to the ranger process, then close STDIN, then 
	# detach and let ranger keep running. 
	gdb -p "$rangerpid"  --batch -ex 'p close(0)' -ex 'detach' -ex 'quit'

	# Now show the CPU usage
	top


Then run:
	
	docker build . -t rtest; docker run --rm -ti --cap-add=SYS_PTRACE --security-opt seccomp=unconfined rtest

This will clone the latest version of ranger and reproduce the 100% cpu usage. 


I have also tried to understand why this only happens some of the time when closing a terminal. The randomness doesn't seem to be due to anything ranger does. Rather, it's a result of the use of Python's `subprocess.Popen()`, because I can easily replicate this without ranger, using plain Python. When starting a new shell and process using `Popen` and then terminate the parent process by closing the terminal, the child process sometimes receives a SIGHUP and terminates, and sometimes continues running. I am not even sure what the intended behavior of `Popen` is in this case.

For ranger, I still think my solution is correct. An alternative solution could be to use a `try-finally`-block and explicitly kill any child processes when ranger exits. But this might not always be the wanted behavior, so I'm not sure this is a good idea.

**Comment by SuperShinyEyes** [comment]
Yes, I followed the instruction given by @igoyak. The issue occurs as described using the given Docker environment. The issue is resolved by his PR.
